File: lib/data/mock/mock_whistleblower_repository.dart
import '../../core/roles.dart';
import '../models/whistleblower_case.dart';
import '../repositories/whistleblower_repository.dart';
import '_latency.dart';
import 'mock_data.dart';

class MockWhistleblowerRepository implements WhistleblowerRepository {
  bool simulateError = false;

  @override
  bool hasAccess(UserRole role) =>
      MockData.whistleblowerAllowedRoles.contains(role);

  @override
  Future<List<WhistleblowerCase>> getCases(UserRole role) async {
    if (!hasAccess(role)) {
      throw WhistleblowerAccessDenied(role);
    }
    await simulateLatency();
    if (simulateError) {
      throw Exception('Mock: failed to load whistleblower cases');
    }
    return MockData.whistleblowerCases();
  }

  @override
  Future<WhistleblowerCase?> getCase(String id, UserRole role) async {
    if (!hasAccess(role)) {
      throw WhistleblowerAccessDenied(role);
    }
    await simulateLatency();
    if (simulateError) {
      throw Exception('Mock: failed to load whistleblower case');
    }
    for (final c in MockData.whistleblowerCases()) {
      if (c.id == id) return c;
    }
    return null;
  }
}

File: Data/Mock/MockWhistleblowerRepository.cs
using OrquestadorFrontend.Core;
using OrquestadorFrontend.Data.Models;
using OrquestadorFrontend.Data.Repositories;

namespace OrquestadorFrontend.Data.Mock;

public sealed class MockWhistleblowerRepository : IWhistleblowerRepository
{
    public bool SimulateError { get; set; }

    public bool HasAccess(UserRole role)
        => MockData.WhistleblowerAllowedRoles.Contains(role);

    public async Task<IReadOnlyList<WhistleblowerCase>> GetCasesAsync(UserRole role)
    {
        // Access check happens before any latency or error simulation so the
        // role gate is enforced even when the demo flips SimulateError on.
        if (!HasAccess(role)) throw new WhistleblowerAccessDeniedException(role);

        await MockLatency.SimulateAsync();
        if (SimulateError) throw new InvalidOperationException("Mock: failed to load whistleblower cases");
        return MockData.WhistleblowerCases();
    }

    public async Task<WhistleblowerCase?> GetCaseAsync(string id, UserRole role)
    {
        if (!HasAccess(role)) throw new WhistleblowerAccessDeniedException(role);

        await MockLatency.SimulateAsync();
        if (SimulateError) throw new InvalidOperationException("Mock: failed to load whistleblower case");
        return MockData.WhistleblowerCases().FirstOrDefault(c => c.Id == id);
    }
}
